File: Payload_Type/thanatos/thanatos/mythic/browser_scripts/netstat.js
function(task, responses){
  if (task.status.includes("error")) {
    return {
      "plaintext": responses.reduce((prev, cur) => prev + cur, ""),
    };
  }

  if (responses.length <= 0) {
    return {"plaintext": "No response yet from agent..."}
  }

  const title = "Network Connections";
  const headers = [
      {"plaintext": "protocol", "type": "string", "cellStyle": {}, "width": 125},
      {"plaintext": "local address", "type": "string", "cellStyle": {}, "width": 300},
      {"plaintext": "local port", "type": "number", "cellStyle": {}, "width": 200},
      {"plaintext": "remote address", "type": "string", "cellStyle": {}, "width": 300},
      {"plaintext": "remote port", "type": "number", "cellStyle": {}, "width": 200},
      {"plaintext": "state", "type": "string", "cellStyle": {}, "width": 200},
      {"plaintext": "pids", "type": "string", "cellStyle": {}, "width": 125},
  ];

  let rows = [];

  for (const response of responses) {
    let data = {};
    try {
      data = JSON.parse(response);
    } catch (error) {
      console.log(error);
      return {
        "plaintext": responses.reduce((prev, cur) => prev + cur, ""),
      }
    }

    rows.push.apply(rows, data.map((netstatEntry) => {
      return {
        "protocol": {
          "plaintext": netstatEntry["proto"],
        },
        "local address": {
          "plaintext": netstatEntry["local_addr"],
          "copyIcon": netstatEntry["local_addr"] !== null && netstatEntry["local_addr"].trim().length !== 0,
        },
        "local port": {
          "plaintext": netstatEntry["local_port"],
          "copyIcon": true,
        },
        "remote address": {
          "plaintext": netstatEntry["remote_addr"],
          "copyIcon": netstatEntry["remote_addr"] !== null && netstatEntry["remote_addr"].trim().length !== 0,
        },
        "remote port": {
          "plaintext": netstatEntry["remote_port"],
          "copyIcon": netstatEntry["remote_port"] !== null && netstatEntry["remote_port"] !== 0,
        },
        "state": {
          "plaintext": netstatEntry["state"],
        },
        "pids": {
          "plaintext": netstatEntry["associated_pids"]?.join(),
          "copyIcon": netstatEntry["associated_pids"] !== null && netstatEntry["associated_pids"].length !== 0,
        },
      }
    }));
  }

  if (rows.length == 0) {
    return {
      "plaintext": responses.reduce((prev, cur) => prev + cur, ""),
    }
  } else {
    return {
      "table": [
        {
          "title": title,
          "headers": headers,
          "rows": rows,
        }
      ]
    };
  }
}

File: Payload_Type/thanatos/thanatos/agent_code/src/netstat.rs
use serde::Serialize;
use crate::agent::AgentTask;
use crate::mythic_success;
use netstat2::{get_sockets_info, AddressFamilyFlags, ProtocolFlags, ProtocolSocketInfo};

/// Struct holding the information for network connections
#[derive(Default, Serialize)]
pub struct NetworkListingEntry {
    /// Protocol
    pub proto: String,

    /// Local address
    pub local_addr: String,

    /// Local Port
    pub local_port: u16,

    /// Remote address
    pub remote_addr: Option<String>,

    /// Remote port
    pub remote_port: Option<u16>,

    /// Associated PIDs
    pub associated_pids: Vec<u32>,

    /// State
    pub state: Option<String>,
}

pub fn netstat(task: &AgentTask) -> Result<(serde_json::Value), Box<dyn std::error::Error>> { 
    let af_flags = AddressFamilyFlags::IPV4 | AddressFamilyFlags::IPV6;
    let proto_flags = ProtocolFlags::TCP | ProtocolFlags::UDP;
    let sockets_info = get_sockets_info(af_flags, proto_flags)?;

    let mut conn: Vec<NetworkListingEntry> = Vec::new();

    for si in sockets_info {
        match si.protocol_socket_info {
            ProtocolSocketInfo::Tcp(tcp_si) => conn.push(NetworkListingEntry {
                proto: "TCP".to_string(),
                local_addr: tcp_si.local_addr.to_string(),
                local_port: tcp_si.local_port,
                remote_addr: Some(tcp_si.remote_addr.to_string()),
                remote_port: Some(tcp_si.remote_port),
                associated_pids: si.associated_pids,
                state: Some(tcp_si.state.to_string()),
            }),
            ProtocolSocketInfo::Udp(udp_si) => conn.push(NetworkListingEntry {
                proto: "UDP".to_string(),
                local_addr: udp_si.local_addr.to_string(),
                local_port: udp_si.local_port,
                remote_addr: None,
                remote_port: None,
                associated_pids: si.associated_pids,
                state: None,
            }),
        }
    }

    let user_output = serde_json::to_string(&conn)?;
    /// Return the output to Mythic
    Ok(mythic_success!(task.id, user_output))
}
